src/Persisters/Flysystem.php

```diff
@@ -66,91 +65,56 @@ class Flysystem implements Persister, Stringable
-     *
-     * @see https://flysystem.thephpleague.com/v1/docs/adapter/ftp/
-     *
-     * @param string $path
-     * @param array<mixed> $config Configuration settings for the FTP adapter.
-     * @param bool $history
-     * @param \Rubix\ML\Persisters\Serializers\Serializer|null $serializer
-     * @return \Rubix\ML\Persisters\Flysystem
-     */
-    public static function ftp(string $path, array $config, bool $history = false, ?Serializer $serializer = null) : Flysystem
-    {
-        $filesystem = new League\Flysystem\Filesystem(new Ftp($config));
-
-        return new Flysystem($path, $filesystem, $history, $serializer);
-    }
-
     /**
      * @param string $path
-     * @param FilesystemInterface $filesystem
+     * @param \League\Flysystem\FilesystemOperator $filesystem
      * @param bool $history
      * @param \Rubix\ML\Persisters\Serializers\Serializer|null $serializer
      */
-    public function __construct(string $path, FilesystemInterface $filesystem, bool $history = false, ?Serializer $serializer = null)
-    {
+    public function __construct(
+        string $path,
+        FilesystemOperator $filesystem,
+        bool $history = false,
+        ?Serializer $serializer = null
+    ) {
         $this->path = $path;
         $this->filesystem = $filesystem;
         $this->history = $history;
         $this->serializer = $serializer ?? new Native();
     }
 
     /**
-     * Save the persistable model.
+     * Save the persistable object.
      *
      * @param \Rubix\ML\Persistable $persistable
      * @throws \RuntimeException
      */
     public function save(Persistable $persistable) : void
     {
-        if ($this->history and $this->filesystem->has($this->path)) {
+        if ($this->history and $this->filesystem->fileExists($this->path)) {
             $timestamp = (string) time();
 
-            $filename = $this->path . '-' . $timestamp . '.' . self::HISTORY_EXT;
+            $filename = "{$this->path}-$timestamp." . self::HISTORY_EXT;
 
             $num = 0;
-            while ($this->filesystem->has($filename)) {
-                $filename = $this->path . '-' . $timestamp . '-' . ++$num . '.' . self::HISTORY_EXT;
+
+            while ($this->filesystem->fileExists($filename)) {
+                $filename = "{$this->path}-$timestamp-" . ++$num . '.' . self::HISTORY_EXT;
             }
 
             try {
-                if (!$this->filesystem->rename($this->path, $filename)) {
-                    throw new RuntimeException("Failed to create history file: '{$filename}' {$this}");
-                }
-            } catch (League\Flysystem\Exception $e) {
-                throw new RuntimeException("Failed to create history file: '{$filename}' {$this}");
+                $this->filesystem->move($this->path, $filename);
+            } catch (FilesystemException $e) {
+                throw new RuntimeException("Failed to create history file '$filename'.");
             }
```

Comment:
I've just noticed that `fileExists` implementations can throw exceptions too. We should moved the `while` statement into the `try` block beneath.

Reply:
Hmmmmm thanks for pointing that